bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/ShellyHandlerFactory.java

```diff
@@ -0,0 +1,161 @@
+    private final Logger logger = LoggerFactory.getLogger(ShellyHandlerFactory.class);
+
+    private final ShellyCoapServer coapServer;
+    private final Set<ShellyDeviceListener> deviceListeners = new CopyOnWriteArraySet<>();
+
+    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = ShellyBindingConstants.SUPPORTED_THING_TYPES_UIDS;
+    private ShellyBindingConfiguration bindingConfig = new ShellyBindingConfiguration();
+
+    /**
+     * Activate the bundle: save properties
+     *
+     * @param componentContext
+     * @param configProperties set of properties from cfg (use same names as in
+     *            thing config)
+     */
+    @Activate
+    public ShellyHandlerFactory(@Reference NetworkAddressService networkAddressService,
+            ComponentContext componentContext, Map<String, Object> configProperties) {
+        logger.debug("Activate Shelly HandlerFactory");
+        super.activate(componentContext);
+
+        this.coapServer = new ShellyCoapServer();
+        Validate.notNull(coapServer, "coapServer creation failed!");
+
+        Validate.notNull(configProperties);
+        bindingConfig.updateFromProperties(configProperties);
+        if (bindingConfig.httpPort == 0) {
+            bindingConfig.httpPort = HttpServiceUtil.getHttpServicePort(componentContext.getBundleContext());
+            if (bindingConfig.httpPort == -1) {
+                bindingConfig.httpPort = 8080;
+            }
+            Validate.isTrue(bindingConfig.httpPort > 0, "Unable to get OH http port");
+            logger.debug("Using OH http port {}", bindingConfig.httpPort);
+        }
+        if (bindingConfig.localIp.isEmpty()) {
+            String lip = networkAddressService.getPrimaryIpv4HostAddress();
+            bindingConfig.localIp = lip != null ? lip : "";
+        }
+    }
+
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        return SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
+    }
+
+    @Override
+    protected @Nullable ThingHandler createHandler(Thing thing) {
+        try {
+            ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+            if (thingTypeUID.getId().equals(THING_TYPE_SHELLYBULB.getId())
+                    || thingTypeUID.getId().equals(THING_TYPE_SHELLYRGBW2_COLOR.getId())
+                    || thingTypeUID.getId().equals(THING_TYPE_SHELLYRGBW2_WHITE.getId())) {
+                logger.debug("Create new thing of type {} using ShellyLightHandler", thingTypeUID.getId());
+                return new ShellyLightHandler(thing, this, bindingConfig, coapServer);
+            } else if (SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID)) {
+                logger.debug("Create new thing of type {} using ShellyRelayHandler", thingTypeUID.getId());
+                return new ShellyRelayHandler(thing, this, bindingConfig, coapServer);
+            }
+        } catch (RuntimeException e) {
```

Comment:
You should not catch RuntimeExceptions, unless you expect a specific exception here and need to catch it.
Otherwise, let it bubble up in the stack, it might e.g. be a out of memory exception or whatsoever.

Reply:
done